src/monattack.cpp

```diff
@@ -425,109 +425,258 @@ void mattack::smash(monster *z, int index)
+ *                      second = the number of empty spaces found.
+ */
+template <size_t N = 1>
+std::pair<std::array<point, (2*N + 1)*(2*N + 1)>, size_t>
+find_empty_neighbors(point const origin) {
+    constexpr auto r = static_cast<int>(N);
+
+    auto const x_min = origin.x - r;
+    auto const x_max = origin.x + r;
+    auto const y_min = origin.y - r;
+    auto const y_max = origin.y + r;
+
+    std::pair<std::array<point, (2*N + 1)*(2*N + 1)>, size_t> result;
+
+    for (auto x = x_min; x <= x_max; ++x) {
+        for (auto y = y_min; y <= y_max; ++y) {
+            if (g->is_empty(x, y)) {
+                result.first[result.second++] = point {x, y};
+            }
+        }
+    }
+
+    return result;
+}
+
+//--------------------------------------------------------------------------------------------------
+/**
+ * Find empty spaces around a creature within a radius of N.
+ *
+ * @see find_empty_neighbors
+ */
+ template <size_t N = 1>
+std::pair<std::array<point, (2*N + 1)*(2*N + 1)>, size_t>
+find_empty_neighbors(Creature const &c) {
+    return find_empty_neighbors<N>(c.pos());
+}
+
+//--------------------------------------------------------------------------------------------------
+/**
+ * Get a size_t value in the closed interval [0, size]; a convenience to avoid messy casting.
+  */
+size_t get_random_index(size_t const size) {
+    return static_cast<size_t>(rng(0, static_cast<long>(size - 1)));
+}
+
+//--------------------------------------------------------------------------------------------------
+/**
+ * Get a size_t value in the closed interval [0, c.size()]; a convenience to avoid messy casting.
+ */
+template <typename Container>
+size_t get_random_index(Container const &c) {
+    return get_random_index(c.size());
+}
+
+void mattack::science(monster *const z, int const index) // I said SCIENCE again!
 {
-    int dist;
-    Creature *target = z->attack_target();
-    if( target == nullptr || 
-        ( dist = rl_dist( z->posx(), z->posy(), target->xpos(), target->ypos() ) ) > 5 || 
-        !z->sees( *target ) ) {
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    // Constants and Configuration
+
+    // attack types
+    enum : int {
+        att_shock,
+        att_radiation,
+        att_manhack,
+        att_acid_pool,
+        att_flavor,
+        att_enum_size
+    };
+
+    // max distance that "science" can be applied to the target.
+    constexpr auto max_distance = 5;
+
+    // attack movement costs
+    constexpr int att_cost_shock   = 0;
+    constexpr int att_cost_rad     = 400;
+    constexpr int att_cost_manhack = 200;
+    constexpr int att_cost_acid    = 100;
+    constexpr int att_cost_flavor  = 80;
+    
+    // radiation attack behaviour
+    constexpr int att_rad_dodge_diff    = 16; // how hard it is to dodge
+    constexpr int att_rad_mutate_chance = 6;  // inverse chance to cause mutation.
```

Comment:
I think I'd like this comment to be more explicit, i.e. `// inverse chance (one in X) to cause mutation`. While the phrasing immediately clicked for me once I read it, I almost skimmed past it, and this doesn't behave like a direct chance (i.e. higher being more likely).